tests/rc_string_ut/rc_string_ut.c

```diff
@@ -26,40 +42,432 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+    THANDLE_DEC_REF(RC_STRING)(rc_string);
+}
+
+/* Tests_SRS_RC_STRING_01_002: [ Otherwise, rc_string_create shall determine the length of string. ]*/
+/* Tests_SRS_RC_STRING_01_003: [ rc_string_create shall allocate memory for the THANDLE(RC_STRING), ensuring all the bytes in string can be copied (including the zero terminator). ]*/
+/* Tests_SRS_RC_STRING_01_004: [ rc_string_create shall copy the string memory (including the NULL terminator). ]*/
+/* Tests_SRS_RC_STRING_01_005: [ rc_string_create shall succeed and return a non-NULL handle. ]*/
+TEST_FUNCTION(rc_string_create_with_empty_string_succeeds)
+{
+    // arrange
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+
+    // act
+    THANDLE(RC_STRING) rc_string = rc_string_create("");
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(char_ptr, "", rc_string->string);
+    ASSERT_IS_NOT_NULL(rc_string);
+
+    // cleanup
+    THANDLE_DEC_REF(RC_STRING)(rc_string);
+}
+
+/* Tests_SRS_RC_STRING_01_006: [ If any error occurs, rc_string_create shall fail and return NULL. ]*/
+TEST_FUNCTION(when_underlying_calls_fail_rc_string_create_also_fails)
+{
+    // arrange
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+
+    umock_c_negative_tests_snapshot();
+
+    for (size_t i = 0; i < umock_c_negative_tests_call_count(); i++)
     {
-        umock_c_deinit();
+        if (umock_c_negative_tests_can_call_fail(i))
+        {
+            umock_c_negative_tests_reset();
+            umock_c_negative_tests_fail_call(i);
 
-        TEST_MUTEX_DESTROY(g_testByTest);
+            // act
+            THANDLE(RC_STRING) rc_string = rc_string_create("gogu");
+
+            ///assert
+            ASSERT_IS_NULL(rc_string, "On failed call %zu", i);
+        }
     }
+}
+
+/* rc_string_create_with_move_memory */
+
+/* Tests_SRS_RC_STRING_01_007: [ If string is NULL, rc_string_create_with_move_memory shall fail and return NULL. ]*/
+TEST_FUNCTION(rc_string_create_with_move_memory_with_NULL_fails)
+{
+    // arrange
+
+    // act
+    THANDLE(RC_STRING) rc_string = rc_string_create_with_move_memory(NULL);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_IS_NULL(rc_string);
+}
+
+/* Tests_SRS_RC_STRING_01_008: [ Otherwise, rc_string_create_with_move_memory shall allocate memory for the THANDLE(RC_STRING). ]*/
+/* Tests_SRS_RC_STRING_01_009: [ rc_string_create_with_move_memory shall associate string with the new handle. ]*/
+/* Tests_SRS_RC_STRING_01_010: [ rc_string_create_with_move_memory shall succeed and return a non-NULL handle. ]*/
+TEST_FUNCTION(rc_string_create_with_move_memory_succeeds)
+{
+    // arrange
+    const char const_test_string[] = "goguletz";
+    char* test_string = (char*)my_gballoc_malloc(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
```

Comment:
also knows as sprintf_char, I guess.

Reply:
Used sprintf_char

---
In reply to: [460261300](https://github.com/Azure/azure-c-util/pull/52#discussion_r460261300) [](ancestors = 460261300)